### DISCUSSION TITLE
Performance, Scalability and SEO

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#3878] [General]
I'm starting to develop an Oqtane application for an international sports federation and I'd like to know if someone has done scalability tests. For example, supporting a few thousand concurrent users in the peaks of maximum use o the average infrastructure requirements in Azure.

Also, I've done a lighthouse report of the https://www.oqtane.org/ site and results can definitely be improved:

![image](https://github.com/oqtane/oqtane.framework/assets/90258222/cb5d2fbd-4c0c-41b1-bc6f-0654790e237d)

Are there any specific guidelines on improving Oqtane performance? I have yet to test the new static version (which would be the ideal setup for my use case) and see if performance improves (which in theory should).

Lastly, afaik Oqtane applications should be able to achieve nice SEO via title and meta tags and crawlable prerendered (in case of server interactivity) and static pages (in case of static ssr). Is this the case?

Thanks for the nice work being done!!

**Comment by sbwalker** [comment]
@mdmontesinos I generally use GTMetrix to do performance testing, which internally uses Lighthouse running on consistent infrastructure to produce consistent results. The Chrome browser plug-in results for Lighthouse can have a huge variance based on the workload of your local machine at any given point in time. The latest GTMetrix report for https://www.oqtane.org is below:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/a2950c55-a7af-42bf-b0c2-abfbe1b7c08d)

The performance score has been higher in the past, however this Azure App Service is now serving 7 websites from the same installation so the load has increased substantially. In addition, the images/content have not been optimized for performance.

Note that https://www.oqtane.org is currently using fully interactive Blazor on .NET 8, as the changes to enable static rendering are still in progress and will be included in the upcoming 5.1 release. I expect the static rendering capability will provide a significant boost in performance as the workload has been significantly optimized.

SEO can definitely be tuned with meta tags, titles, etc... This was already possible with interactive rendering and is also supported in static rendering.

**Comment by sbwalker** [comment]
Significant improvement with Oqtane 5.1 using Static Rendering (and no other optimizations to web content)


![image](https://github.com/oqtane/oqtane.framework/assets/4840590/0b6ebfe1-6296-4b7e-9984-265762acbebf)

**Comment by sbwalker** [comment]
I just realized that the Contact form at the bottom of the Home page was being rendered interactively... so the performance measurement above was not based on a pure statically rendered page. I converted the Contact form to static rendering and the result improved slightly:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/ba49ceca-36e8-4813-aca9-8f7e183d2d26)
 
Note that this 2% improvement is simply by removing a single interactive component from the page. I think this is a good example of how static rendering can offer a significant improvement for both performance and scalability.

**Comment by mdmontesinos** [comment]
Thanks for the updates @sbwalker! In the following months I'll be developing a considerably sized project using Oqtane in static render mode, so performance and scalability will be critical. The new 5.1 release looks amazing in terms of both... When the project is in a more advanced state I'll try to post some data on stress testing.

**Comment by sbwalker** [comment]
Updated my Azure App Service to .NET SDK 8.0.4 and 64 bit processing... and performance improved slightly:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/4936f641-cc59-4a78-9a65-75206f7a625a)

**Comment by sbwalker** [comment]
And now with Oqtane 6.1 on .NET 9 with Static Asset Caching enabled:

![image](https://github.com/user-attachments/assets/8bb73423-3f3e-4c3e-8273-8d94ceee81e9)

**Comment by mdmontesinos** [comment]
Oqtane 6.1 with Static Render Mode and also static asset caching enabled:

![{E672974E-E240-4847-9C17-87FF5A8E50EA}](https://github.com/user-attachments/assets/1dc5008f-8c4d-4836-8d65-b9448ed46c44)

And also from Google Lighthouse:

![{7696FF7D-8135-46AB-B8A5-37702D56C1AD}](https://github.com/user-attachments/assets/220f2882-e7ad-439b-a815-e3e53a51f52a)

Note that this is the home page of the site, and contains around 10 custom modules, with loads of requests for external data and images. Everything is optimized as much as possible for performance.

Definitely Oqtane can be used to build really performant and SEO-optimized sites!!
